File: config/src/main/scala/net/kemitix/thorp/config/ConfigurationBuilder.scala
package net.kemitix.thorp.config

import java.io.File

import zio.Task

/**
  * Builds a configuration from settings in a file within the
  * `source` directory and from supplied configuration options.
  */
trait ConfigurationBuilder {

  private val userConfigFile = ".config/thorp.conf"
  private val globalConfig   = new File("/etc/thorp.conf")
  private val userHome       = new File(System.getProperty("user.home"))

  def buildConfig(priorityOpts: ConfigOptions): Task[Configuration] =
    (getConfigOptions(priorityOpts).map(collateOptions) >>=
      ConfigValidator.validateConfig)

  private def getConfigOptions(
      priorityOpts: ConfigOptions): Task[ConfigOptions] = {
    val sourceOpts =
      SourceConfigLoader.loadSourceConfigs(ConfigQuery.sources(priorityOpts))
    val userOpts   = userOptions(priorityOpts ++ sourceOpts)
    val globalOpts = globalOptions(priorityOpts ++ sourceOpts ++ userOpts)
    Task(priorityOpts ++ sourceOpts ++ userOpts ++ globalOpts)
  }

  private def userOptions(priorityOpts: ConfigOptions): ConfigOptions =
    if (ConfigQuery.ignoreUserOptions(priorityOpts))
      ConfigOptions.empty
    else ParseConfigFile.parseFile(new File(userHome, userConfigFile))

  private def globalOptions(priorityOpts: ConfigOptions): ConfigOptions =
    if (ConfigQuery.ignoreGlobalOptions(priorityOpts))
      ConfigOptions.empty
    else ParseConfigFile.parseFile(globalConfig)

  private def collateOptions(configOptions: ConfigOptions): Configuration =
    ConfigOptions.options
      .get(configOptions)
      .foldLeft(Configuration.empty) { (config, configOption) =>
        configOption.update(config)
      }

}

object ConfigurationBuilder extends ConfigurationBuilder

File: config/src/main/java/net/kemitix/thorp/config/ConfigurationBuilder.java
package net.kemitix.thorp.config;

import java.io.File;
import java.io.IOException;

public interface ConfigurationBuilder {
    static Configuration buildConfig(ConfigOptions priorityOpts) throws IOException, ConfigValidationException {
        return new ConfigurationBuilderImpl().buildConfig(priorityOpts);
    }
    class ConfigurationBuilderImpl implements ConfigurationBuilder {
        private static final String userConfigFile = ".config/thorp.conf";
        private static final File globalConfig = new File("/etc/thorp.conf");
        private static final File userHome = new File(System.getProperty("user.home"));
        Configuration buildConfig(ConfigOptions priorityOpts) throws IOException, ConfigValidationException {
            return ConfigValidator.validateConfig(
                    collateOptions(getConfigOptions(priorityOpts)));
        }
        private ConfigOptions getConfigOptions(ConfigOptions priorityOpts) throws IOException {
            ConfigOptions sourceOpts = SourceConfigLoader.loadSourceConfigs(ConfigQuery.sources(priorityOpts));
            ConfigOptions userOpts = userOptions(priorityOpts.merge(sourceOpts));
            ConfigOptions globalOpts = globalOptions(priorityOpts.merge(sourceOpts.merge(userOpts)));
            return priorityOpts.merge(sourceOpts.merge(userOpts.merge(globalOpts)));
        }
        private ConfigOptions userOptions(ConfigOptions priorityOpts) throws IOException {
            if (ConfigQuery.ignoreUserOptions(priorityOpts)) {
                return ConfigOptions.empty();
            }
            return ParseConfigFile.parseFile(
                    new File(userHome, userConfigFile));
        }
        private ConfigOptions globalOptions(ConfigOptions priorityOpts) throws IOException {
            if (ConfigQuery.ignoreGlobalOptions(priorityOpts)) {
                return ConfigOptions.empty();
            }
            return ParseConfigFile.parseFile(globalConfig);
        }
        private Configuration collateOptions(ConfigOptions configOptions) {
            Configuration config = Configuration.create();
            for (ConfigOption configOption : configOptions.options()) {
                config = configOption.update(config);
            }
            return config;
        }
    }
}
